pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -604,6 +632,64 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 			// If the container failed the startup probe, we should kill it.
 			message = fmt.Sprintf("Container %s failed startup probe", container.Name)
 		} else {
+			if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+				keepCount++
+				if container.Resources.Limits == nil || len(pod.Status.ContainerStatuses) == 0 {
+					continue
+				}
+				apiContainerStatus, exists := podutil.GetContainerStatus(pod.Status.ContainerStatuses, container.Name)
```

Comment:
ideally, we shouldn't depend on the API pod status here.  We should just reconcile the observed (kubecontainer) container status to the resources allocated.

Reply:
this is to support the transition period where runtime has not implemented reporting Resources in ContainerStatus API response.

apiContainerStatus is populated based on the Spec resources upon successful container creation (initial pod running), and if we get this from CRI, that supercedes spec info. For resizes, we use spec info only if CRI does not yet support it, and updateContainerResources was successful.

Reply:
So essentially you are saying: `v1 pod spec + kubecontainer status = v1 pod status`?  The main thing I am concerned about is that pod.Status.Container[i]... is usually the one stored in the API Server, which can be stale.  I'd prefer only depending on pod.Spec and kubecontainer status if possible.  Could we instead make a helper function for some of the api status generation logic here?  Or we could explictly regenerate the "new" api status to ensure we have consistent behavior.

Reply:
I had to do this because kubeContainerStatus may be insufficient, and pod.Spec cannot be used without caveats. There are two cases that we have to handle:
1. Runtime has been updated for CRI API change, and returns latest cgroup config in ContainerStatus.Resources,
1. Runtime has not been yet been updated to populate Resources.

Case 1 is easy, we always use cgroup config returned by runtime when we call GetPodStatus after UpdateContainerResources (for newly started container, PLEG is already calling GetPodStatus to query runtime). In this case, kubelet_pods uses runtime supplied config in kubeContainerStatus.Resources to generate apiContainerStatus.

Case 2 is where things get interesting, because we get nothing back from runtime. So we have to go with what we last know - which is what we put into apiContainerStatus. 
- For new container. the container-state transition to running means new container was started, and we trust that the cgroup config passed to the runtime in startContainer has been applied, and we use Resources from Spec.
- For resize, when updateContainerResources is successful, we trust that runtime applied the cgroup config we passed in update, and set that info in pod-cache, which gets consumed when apiStatus is next generated.

So, if we manage what we put into apiContainerStatuses carefully, then we should be ok and I documented in the comments where apiStatus is updated in [kubelet_pods.go](https://github.com/kubernetes/kubernetes/pull/92127/files#diff-10055ae93a8699af13ceba0482fc43c3R1579) 

Now, if the information in apiContainerStatus does get stale or is somehow lost, what's the worst that can happen? We end up executing the update code, and will result in an unnecessary call to updateContainerResources telling runtime to apply config that has already been applied. At some point, we should get rid of case 2 (not sure about the process - perhaps force runtimes to update by adding a gating test if we have such a thing?).

Reply:
The CRI API is supposed to require a set of behaviors of container runtimes.  I agree that we shouldn't require them to add the new function _immediately_, but we should require it eventually, even if they just return exactly what we created the container with...

I wonder if just using the in-place-update feature-gate is the right mechanism?  I.E. Alpha is the opt-in period for runtimes, Beta is when most should have it implemented, but can opt-out if they haven't, and GA is when it is required?  It feels like we are re-implementing the feature graduation process with fall-backs...

Reply:
Feature-gate alone may not be sufficient to raise awareness. I have to do some research on the process, but I'm hoping that besides socializing the updated ContainerStatus API and updateContainerResources APIs with runtime maintainers, we have some tests that runtimes have to pass in order to work with changes like this.
I think it is reasonable to give at-least one release-cycle after GA before such test becomes mandatory. That should give runtimes enough runway to plan, implement, test ..

Reply:
Ok.  I'm fine with the current implementation for now.  Maybe as a follow-up, you can try and make the 'code to be removed' easier to rip out at a later date, since it is always hard to tell what is safe to remove vs dangerous in later releases.

Reply:
I'll work on a patch and test it out, and create an issue to track it and apply the patch post-GA of this feature. Let's iron out the details of the plan tomorrow.